mesonpy/_tags.py

```diff
@@ -1,130 +1,124 @@
-        self._abi_number = int(match.group('abi_number'))
-
-    @property
-    def abi_number(self) -> int:
-        return self._abi_number
-
-    def __str__(self) -> str:
-        return f'abi{self.abi_number}'
-
-    @property
-    def python(self) -> Literal[None]:
-        return None
-
-    @property
-    def abi(self) -> str:
-        return f'abi{self.abi_number}'
-
-    def __eq__(self, other: Any) -> bool:
-        return isinstance(other, self.__class__) and other.abi_number == self.abi_number
-
-    def __hash__(self) -> int:
-        return hash(str(self))
-
-
-class InterpreterTag(Tag):
-    def __init__(self, value: str) -> None:
-        parts = value.split('-')
-        if len(parts) < 2:
-            raise ValueError(
-                'Invalid PEP 3149 interpreter tag, expected at '
-                f'least 2 parts but got {len(parts)}'
-            )
-
-        # On Windows, file extensions look like `.cp311-win_amd64.pyd`, so the
-        # implementation part (`cpython-`) is different from Linux. Handle that here:
-        if parts[0].startswith('cp3'):
-            parts.insert(0, 'cpython')
-            parts[1] = parts[1][2:]  # strip 'cp'
-
-        self._implementation = parts[0]
-        self._interpreter_version = parts[1]
-        self._additional_information = parts[2:]
-
-        if self.implementation != 'cpython' and not self.implementation.startswith('pypy'):
-            raise NotImplementedError(f'Unknown Python implementation: {self.implementation}.')
-
-    @property
-    def implementation(self) -> str:
-        return self._implementation
-
-    @property
-    def interpreter_version(self) -> str:
-        return self._interpreter_version
-
-    @property
-    def additional_information(self) -> Sequence[str]:
-        return tuple(self._additional_information)
+import os
+import platform
+import sys
+import sysconfig
+
+from typing import Optional, Union
+
+
+# https://peps.python.org/pep-0425/#python-tag
+INTERPRETERS = {
+    'python': 'py',
+    'cpython': 'cp',
+    'pypy': 'pp',
+    'ironpython': 'ip',
+    'jython': 'jy',
+}
+
+
+def get_interpreter_tag() -> str:
+    name = sys.implementation.name
+    name = INTERPRETERS.get(name, name)
+    version = sys.version_info
+    return f'{name}{version[0]}{version[1]}'
+
+
+def _get_config_var(name: str, default: Union[str, int, None] = None) -> Union[str, int, None]:
+    value = sysconfig.get_config_var(name)
+    if value is None:
+        return default
+    return value
+
+
+def _get_cpython_abi() -> str:
+    version = sys.version_info
+    debug = pymalloc = ''
+    if _get_config_var('Py_DEBUG', hasattr(sys, 'gettotalrefcount')):
+        debug = 'd'
+    if sys.version_info < (3, 8) and _get_config_var('WITH_PYMALLOC', True):
+        pymalloc = 'm'
+    return f'cp{version[0]}{version[1]}{debug}{pymalloc}'
+
+
+def get_abi_tag() -> str:
+    abi = sysconfig.get_config_var('SOABI')
```

Comment:
The `SOABI` isn't documented anywhere, so could be removed in future Python versions. While I am okay merging this as an initial implementation, granted that all of our current tests pass, we should work on figuring out a better way to do this.

Please add an `# XXX: ...` comment documenting this.

Reply:
`packaging` relies on the same configuration variable. As we need to be bug-compatible with `packaging` I figure this is okay. AFAIK, none of the variables available with `sysconfig.get_config_var()` is documented. Should I add the same comment everywhere we use one?

Reply:
It does, but only for the non CPython tags. The normalized version of `SOABI` is used as the full ABI tag as a fallback, we can do the same.

IIRC we only use `SOABI` in one other place, and it has a fallback value.

Reply:
We use other variables from `sysconfig.get_config_var()` all of them are undocumented. Aren't they?

Reply:
It is documented in [PEP 3149](https://peps.python.org/pep-3149/) and should be available on all CPythons, along with `EXT_SUFFIX`. As I mentioned on the other thread, `EXT_SUFFIX` is more reliable due to a long standing bug in packaging expecting a broken `SOABI` because PyPy has a broken SOABI. You can get a little backstory in https://foss.heptapod.net/pypy/pypy/-/issues/3816 or https://gitlab.kitware.com/cmake/cmake/-/issues/24121.